File: problems/CSES/Graph_Algorithms/round_trip/round_trip.py
import sys
sys.setrecursionlimit(300000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n + 1)]

    for _ in range(m):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    visited = [False] * (n + 1)
    parent = [-1] * (n + 1)

    cycle_start = -1
    cycle_end = -1

    def dfs(u, p):
        nonlocal cycle_start, cycle_end

        visited[u] = True

        for v in graph[u]:
            if v == p:
                continue

            if visited[v]:
                cycle_start = v
                cycle_end = u
                return True

            parent[v] = u

            if dfs(v, u):
                return True

        return False

    for i in range(1, n + 1):
        if not visited[i]:
            if dfs(i, -1):
                break

    if cycle_start == -1:
        print("IMPOSSIBLE")
        return

    cycle = [cycle_start]

    cur = cycle_end

    while cur != cycle_start:
        cycle.append(cur)
        cur = parent[cur]

    cycle.append(cycle_start)

    print(len(cycle))
    print(*cycle)

main()

File: problems/CSES/Graph_Algorithms/round_trip/round_trip.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int n, m;

vector<vector<int>> graph;
vector<bool> visited;
vector<int> parent;

int cycle_start = -1;
int cycle_end = -1;

bool dfs(int u, int p) {
    visited[u] = true;

    for (int v : graph[u]) {
        if (v == p)
            continue;

        if (visited[v]) {
            cycle_start = v;
            cycle_end = u;
            return true;
        }

        parent[v] = u;

        if (dfs(v, u))
            return true;
    }

    return false;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> m;

    graph.resize(n + 1);
    visited.assign(n + 1, false);
    parent.assign(n + 1, -1);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;

        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            if (dfs(i, -1))
                break;
        }
    }

    if (cycle_start == -1) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }

    vector<int> cycle;

    cycle.push_back(cycle_start);

    int cur = cycle_end;

    while (cur != cycle_start) {
        cycle.push_back(cur);
        cur = parent[cur];
    }

    cycle.push_back(cycle_start);

    cout << cycle.size() << '\n';

    for (int node : cycle) {
        cout << node << ' ';
    }

    cout << '\n';

    return 0;
}
